### PR TITLE
[Torch] add torch overview

### PR DESCRIPTION
**Pull request by fortianyou**

### CODE REVIEW

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+## Overview
+
+![bladedisc_torch_overview](./bladedisc_torch_overview.jpg)
+
+The figure shows how a PyTorch nn.Module is compiled via BladeDISC. In short,
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:5]: Shall we setup a convention that anything referring to variable/function/class name in code should be quoted by \` notation ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+BladeDISC for PyTorch is a compiler to python subset language for deep learning.
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:6]: ... a compiler targeting a subset of Python language for ... ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+A PyTorch nn.Module's forward mehtod is compiled to TorchScript by scripting or
+tracing firstly. Later, some passes are runned to simplify the TorchScript graph
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:9]: runned -> run

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+more closer to SSA(Static Single Assignment) Tensor computations. Then, some
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:10]: more closer -> closer ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+lowering from TorchScript to MHLO
+
+- Query if the conversion of a graph node would succeed under a certain
+  conversion context
+- Build an MHLO module during converting TorchScript subgraph to MHLO
+
+### 4. Clustering and fallback
+
+TorchScript language is a subset of python for Deep Learning. There are many
+primitive operators and data structures that are reflected by python. MHLO can
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:53]: There are -> It has ..
... data structures ~that are~ reflected by python.

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+only support a subset of operators of TorchScript where tensors are values that
+can't be modified. Those issues are called *language subset problems*. BladeDISC
+addresses such problems with a clustering and fallback mechanism:
+
+- Iteratively, clean pythonic IR's and reduce the TorchScript graph closer to
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:58]: clean pythonic IR ~'s~ and ... ? 
What do you mean by '**reduce** the TorchScript graph **closer** to ...' ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  computations with SSA semantics
+- Collect supported candidate nodes to backend with the simplified graph
+- Do clustering on the supported subset of nodes; The cluster had to be canceled
+  if a cyclic was constructed
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:62]: cyclic -> cycle ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+- Finally, fuse the clustered subgraph and lower it to the compiler backend; All
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:63]: nit: the compiler backend -> DISC ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  the nodes outside the clusters are fallback to TorchScript.
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:64]: are fallback -> fall back ?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+### 5. Runtime(RAL, Runtime Abstraction Layer)
+
+The runtime of our BladeDISC compiler, [more details](<>)
+
+## Some additional views
+
+### 1. Relations between aten::Tensor and mlir::Value(Tensor)
+
+An `aten::Tensor` actually is a memory buffer that are shaped, so that it could
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:74]: is a shaped memory buffer that could be viewed, sliced and modified in place?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+be viewed, modified in place and slices. However, `mlir::Value` could not be
+modified. Any operations on `mlir::Value` would create new `mlir::Value`. The
+memory buffer was neglected until graph was lowering to LMHLO.
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:77]: the word neglected may not be correct, IMO. In fact it does **not EXIST**, not be **neglected**.

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+### 2. Relations between Scalar and 0-rank Tensor
+
+- A torch aten tensor can be viewed, reshaped, sliced and modified. But a torch
+  Scalar can't.
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:82]: Need this period in the end? while the others don't have.

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+- A 0-rank tensor in MHLO is equivalent to a scalar, and they could be converted
+  to each other
+- In BladeDISC, torch Scalar was converted to mlir standard dialect operator
+  mlir::ConstantOp
+
+### 3. How BladeDISC converts List\[Scalar\] and List\[Tensor\]
+
+BladeDISC only converts those Lists of Scalars and Tensors that whose elements
```
Comment (qiuxiafei) [inline pytorch_addons/docs/bladedisc_torch_overview.md:90]: BladeDISC only converts ~those~ Lists of Scalars and Tensors ~that~ whose elements

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 1 above)
Comment (yangjunpro) [reply pytorch_addons/docs/bladedisc_torch_overview.md:5]: +1

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 2 above)
Comment (yangjunpro) [reply pytorch_addons/docs/bladedisc_torch_overview.md:6]: "a compiler to python subset language for deep learning" I am also puzzled about the meaning of this sentence:)

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 14]
```diff
@@ -0,0 +1,112 @@
+## Overview
+
+![bladedisc_torch_overview](./bladedisc_torch_overview.jpg)
+
+The figure shows how a PyTorch nn.Module is compiled via BladeDISC. In short,
+BladeDISC for PyTorch is a compiler to python subset language for deep learning.
+
+A PyTorch nn.Module's forward mehtod is compiled to TorchScript by scripting or
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:8]: Nit."mehtod"==>"method".

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 3 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:9]: Nit."are runned"==>"are run"

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 4 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:10]: Nit."more closer "==>"to make it closer"?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+tracing firstly. Later, some passes are runned to simplify the TorchScript graph
+more closer to SSA(Static Single Assignment) Tensor computations. Then, some
+subgraphs are converted to MHLO. Once BladeDISC get a MHLO module, it will
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:11]: Nit."get"==>“gets”.

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+compiler world, with the considerations:
+
+- Dynamic shape support
+- Control flow support
+- Available and well established BladeDISC compiler
+- Continuous improvements and updates from the MLIR/XLA community
+
+### 2. Conversion & Lowering
+
+During optimization and compilation, there are many conversions and lowering
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:31]: Nit."there are many ... operators"==>"TorchScript will be converted and lowered to MHLO operators with the following work:

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+among TorchScript and MHLO operators. We had to do the following things:
+
+- Inline the functions and methods
+- Freeze the parameters while reserving the required attributes
+- Reduce and concrete the ranks and data types of the tensor and parameters
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:36]: what does "concrete" mean here?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+- Available and well established BladeDISC compiler
+- Continuous improvements and updates from the MLIR/XLA community
+
+### 2. Conversion & Lowering
+
+During optimization and compilation, there are many conversions and lowering
+among TorchScript and MHLO operators. We had to do the following things:
+
+- Inline the functions and methods
+- Freeze the parameters while reserving the required attributes
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:35]: Nit."reserving"==>"preserving"?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+- Reduce and concrete the ranks and data types of the tensor and parameters
+- Do constant folding and remove irrelevance or dead codes
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:37]: Nit."irrelevant".

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+- Do aliasing analysis and rewrite in-place operations to out-of-place
+  corresponding versions
+
+### 3. Converters
+
+The IR converters are exactly the main mechanism that supports converting and
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:43]: "that supports converting and
lowering from TorchScript to MHLO"==>"to lower TorchScript to MHLO" may be more compact?

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 5 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:53]: What does "reflected" mean here...

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+
+- Query if the conversion of a graph node would succeed under a certain
+  conversion context
+- Build an MHLO module during converting TorchScript subgraph to MHLO
+
+### 4. Clustering and fallback
+
+TorchScript language is a subset of python for Deep Learning. There are many
+primitive operators and data structures that are reflected by python. MHLO can
+only support a subset of operators of TorchScript where tensors are values that
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:54]: "support"==>"supports"

Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:54]: "where tensors are value that can't be modified"==>"where tensors are immutable values"

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 6 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:58]: "reduce"==>"simplify"?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+can't be modified. Those issues are called *language subset problems*. BladeDISC
+addresses such problems with a clustering and fallback mechanism:
+
+- Iteratively, clean pythonic IR's and reduce the TorchScript graph closer to
+  computations with SSA semantics
+- Collect supported candidate nodes to backend with the simplified graph
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:60]: What does "to backend" mean here?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+- Do clustering on the supported subset of nodes; The cluster had to be canceled
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:61]: "The cluster"==>"The clustering".

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 7 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:62]: "was constructed"==>"introduce a cycle"?

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 9 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:64]: "outside of"？

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 10 above)
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:74]: What specifically does "are shaped" mean here?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+### 5. Runtime(RAL, Runtime Abstraction Layer)
+
+The runtime of our BladeDISC compiler, [more details](<>)
+
+## Some additional views
+
+### 1. Relations between aten::Tensor and mlir::Value(Tensor)
+
+An `aten::Tensor` actually is a memory buffer that are shaped, so that it could
+be viewed, modified in place and slices. However, `mlir::Value` could not be
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:75]: "slices“==>"sliced”

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+modified. Any operations on `mlir::Value` would create new `mlir::Value`. The
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:76]: “could not be modified due to its SSA nature".

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+- A 0-rank tensor in MHLO is equivalent to a scalar, and they could be converted
+  to each other
+- In BladeDISC, torch Scalar was converted to mlir standard dialect operator
+  mlir::ConstantOp
+
+### 3. How BladeDISC converts List\[Scalar\] and List\[Tensor\]
+
+BladeDISC only converts those Lists of Scalars and Tensors that whose elements
+could be analyzed statically at conversion time. Once the elements are known, we
+could imitate the operations of List at conversion time, so that there is no
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:92]: What does "imitate" mean here?

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+need to preserve List any more. Otherwise, BladeDISC would fallback the List
+operations to TorchScript/PyTorch. So dose Dict.
```
Comment (yangjunpro) [inline pytorch_addons/docs/bladedisc_torch_overview.md:94]: "dose"==>"does"

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+
+A PyTorch nn.Module's forward mehtod is compiled to TorchScript by scripting or
+tracing firstly. Later, some passes are runned to simplify the TorchScript graph
+more closer to SSA(Static Single Assignment) Tensor computations. Then, some
+subgraphs are converted to MHLO. Once BladeDISC get a MHLO module, it will
+optimize and lower the module progressively to different backends.
+
+BladeDISC for PyTorch only supports AOT compilation currently. While JIT's
+roadmap should not differ too much. Please go to
+[Pass Pipeline Walkthrough](../../docs/blade_disc_pass_pipeline_walkthrough.md)
```
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:16]: pls redirect the link to ../../docs/pass_pipeline.md

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+
+- Dynamic shape support
+- Control flow support
+- Available and well established BladeDISC compiler
+- Continuous improvements and updates from the MLIR/XLA community
+
+### 2. Conversion & Lowering
+
+During optimization and compilation, there are many conversions and lowering
+among TorchScript and MHLO operators. We had to do the following things:
```
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:32]: had -> have
or, "The following things are needed to be done:"

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+  computations with SSA semantics
+- Collect supported candidate nodes to backend with the simplified graph
+- Do clustering on the supported subset of nodes; The cluster had to be canceled
+  if a cyclic was constructed
+- Finally, fuse the clustered subgraph and lower it to the compiler backend; All
+  the nodes outside the clusters are fallback to TorchScript.
+
+### 5. Runtime(RAL, Runtime Abstraction Layer)
+
+The runtime of our BladeDISC compiler, [more details](<>)
```
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:68]: we have the link now

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+- Do constant folding and remove irrelevance or dead codes
+- Do aliasing analysis and rewrite in-place operations to out-of-place
+  corresponding versions
+
+### 3. Converters
+
+The IR converters are exactly the main mechanism that supports converting and
+lowering from TorchScript to MHLO
+
+- Query if the conversion of a graph node would succeed under a certain
```
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:46]: would -> can

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 9 above)
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:64]: outside -> outside of

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 32]
```diff
@@ -0,0 +1,112 @@
...
+  Scalar can't.
+- A 0-rank tensor in MHLO is equivalent to a scalar, and they could be converted
+  to each other
+- In BladeDISC, torch Scalar was converted to mlir standard dialect operator
+  mlir::ConstantOp
+
+### 3. How BladeDISC converts List\[Scalar\] and List\[Tensor\]
+
+BladeDISC only converts those Lists of Scalars and Tensors that whose elements
+could be analyzed statically at conversion time. Once the elements are known, we
```
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:91]: could -> can

File: pytorch_addons/docs/bladedisc_torch_overview.md (hunk 27 above)
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:94]: ... Pytorch, so does Dict.
comma should be used here.

File: pytorch_addons/docs/bladedisc_torch_overview.md [hunk 33]
```diff
@@ -0,0 +1,112 @@
...
+need to preserve List any more. Otherwise, BladeDISC would fallback the List
+operations to TorchScript/PyTorch. So dose Dict.
+
+## Difference to Torch-MLIR
+
+Both [TorchMLIR](https://github.com/llvm/torch-mlir) and BladeDISC aims to
+leverage MLIR as the middle layer mapping from PyTorch and progressively lower
+down to different target hardwares.
+
+The Torch-MLIR project aims to provide first class compiler support from the
```
Comment (linearhit) [inline pytorch_addons/docs/bladedisc_torch_overview.md:102]: nit: first class -> robust and smooth

**Pull request closed without merging**